dspy/adapters/base.py

```diff
@@ -32,3 +40,14 @@ def __call__(self, lm, lm_kwargs, signature, demos, inputs, _parse_values=True):
                 return JsonAdapter()(lm, lm_kwargs, signature, demos, inputs, _parse_values=_parse_values)
             raise e
 
+    @abstractmethod
+    def format(self, signature, demos, inputs):
+       raise NotImplementedError
+
+    @abstractmethod
+    def format_finetune_data(self, signature, demos, inputs, outputs):
```

Comment:
yes

Reply:
Sg -- a note: the first time we check whether a model can be fine-tuned is `lm.finetune(...)`, but we call this function before that in `BootstrapFinetune` -- this function is just meant to put together the prompt along with completion (e.g. in a message format, plain format) corresponding to how the adapter formats these. It can be used by any process that needs to have access to the full call (e.g. could be fine-tuning, other optimizers) We could perhaps re-name.